### PR TITLE
Send and estimate gas linear fit function

### PR DESCRIPTION
**Pull request by wilsoniovlabs**
## What

- Adding estimate max possible relay gas relay with linear fit function
- Adding relay a send custom transaction

## Why

- Implement execute modal on DAPP for SDK.

## Refs

- (optional: include links to other issues, PRs, tickets, etc)

### CODE REVIEW

**Review by antomor — CHANGES_REQUESTED** [review]
Apart from some minor typos and a couple of doubts I have, it looks good to me!

File: src/index.ts [hunk 1]
```diff
@@ -105,13 +105,46 @@ interface RelayingServices {
     /**
      * It executes a estimate max possible relay gas to get a number value
      *
-     * @param trxDetails the transaction details
+     * @param smartWallet the transaction details
```
[the remark this answers is not in the document]
Comment (mortelli) [reply src/index.ts:108]: agreed, although i don't understand the original line either

**Review by mortelli — COMMENTED** [review]
before i fully review this, i have 2 questions:

1. what is the difference between `relayTransaction` and `send`? they appear to serve the same purpose but with different signatures. couldn't one call the other?
2. why do we need to estimate the gas linearly for the execute feature? is there something wrong with using the old estimation?

thanks

**Review by antomor — CHANGES_REQUESTED** [review]
Apart from the unresolved comments, the execution of the tests is failing.

**Comment by wilsoniovlabs** [comment]
> before i fully review this, i have 2 questions:
> 
> 1. what is the difference between `relayTransaction` and `send`? they appear to serve the same purpose but with different signatures. couldn't one call the other?
> 2. why do we need to estimate the gas linearly for the execute feature? is there something wrong with using the old estimation?
> 
> thanks

1) This function was removed.
2) The provider need to have 2 way to estimate the gas, fist one with validation on the walles and the secound to validate without using the wallet.

**Review by mortelli — COMMENTED** [review]
@antomor tests appear to be passing on my end, can you please detail what error you are getting?

thanks

**Comment by antomor** [comment]
> @antomor tests appear to be passing on my end, can you please detail what error you are getting?
> 
> thanks

They are passing now, because of the commit [48050a8](https://github.com/infuy/relaying-services-sdk/pull/9/commits/48050a8062b137b92c7a99585bc21c84acbfde38)

File: src/sdk.ts [hunk 2]
```diff
@@ -440,7 +440,49 @@ export class DefaultRelayingServices implements RelayingServices {
...
+            await this.relayProvider.relayClient.estimateMaxPossibleRelayGasWithLinearFit(
+                trxDetails,
+                relayWorker
+            );
+        const maxPossibleGas = toBN(maxPossibleGasValue);
+        const estimate = maxPossibleGas.mul(gasPrice);
+        return Number(estimate.toString());
+    }
+
+    private _getAccountAddress(): string {
```
Comment (mortelli) [inline src/sdk.ts:485]: is there any reason why this function was renamed?

**Review by mortelli — COMMENTED** [review]
some comments

File: src/sdk.ts (hunk 2 above)
Comment (wilsoniovlabs) [reply src/sdk.ts:485]: getAccountAddress was rename to avoiding confusion because it is a private method.

Comment (mortelli) [reply src/sdk.ts:485]: i don't think it was necessary, but it's not a blocker

(see [here](https://stackoverflow.com/questions/47747809/typescript-underscore-convention-for-members))

File: src/sdk.ts [hunk 3]
```diff
@@ -440,7 +440,110 @@ export class DefaultRelayingServices implements RelayingServices {
         return estimate.toString();
     }
 
-    getAccountAddress(): string {
+    async estimateMaxPossibleRelayGasWithLinearFit(
```
[the remark this answers is not in the document]
Comment (wilsoniovlabs) [reply src/sdk.ts:443]: The difference  is the position change on the conde, by in fact all is the same.

[the remark this answers is not in the document]
Comment (antomor) [reply src/sdk.ts:443]: Ok, the only concerns I have is related to the return type: are we sure that the `BN` can be converted into a `Number`? IMHO I wouldn't do this operation in the sdk

[the remark this answers is not in the document]
Comment (wilsoniovlabs) [reply src/sdk.ts:443]: I Agreed with that, i change it to return string because the other one is returning a string.

**Review by antomor — APPROVED** [review]

**Review by mortelli — APPROVED** [review]
thanks

**Pull request merged**